### DISCUSSION TITLE
email or username login

### DISCUSSION THREAD

**Discussion by deelight-del** [deelight-del/joke-witter#13] [General]
I am still trying to wrap my head around using email or username to log in.

I am considering an edge case where two different users have coinciding email and username. For instance, UserA has the email `foo@mail.com` and UserB has the username `foo@mail.com`. When they want to log in, the integrity of the DB might be compromised.

To mitigate this, I think we should disallow the at symbol (\@) in a username when registering. This way, we can be sure that entries with the \@ symbol are emails and otherwise usernames.
@Ayobami0

**Comment by Ayobami0** [comment]
That's true. Yeah that should work

**Reply by Ayobami0** [reply]
There would be a need to validate either email address or username then. Or will it be the user that would insert the @ symbol

**Reply by deelight-del** [reply]
Yeah, there will be a need to validate the email and username.

What I think about the validation is to disallow the use of \@ when the user enters their username. If the username they are registering with contains an \@ symbol, then we can raise a bad request back to the front end. On the other hand, any entered mail should contain an \@ symbol.

Just as a sidenote, the new PR that I will send will change the email to an `EmailField`, which I think should have additional validation abilities for email validation. I haven't tested it though.

I am not sure I got the second part - 'Or will it be the user that would insert the @ symbol'

**Reply by Ayobami0** [reply]
Yeah, would the user include the @ when signing in so something like @myusername or just myusername

**Reply by deelight-del** [reply]
I think just myusername is fine...

**Comment by Ayobami0** [comment]
Ok, the adding will be done when storing and retrieving

**Reply by deelight-del** [reply]
Oooh... I'm not sure I understand why we need the @ symbol for the username in the dB...

**Reply by Ayobami0** [reply]
🤦

**Reply by Ayobami0** [reply]
I miss understood

**Reply by Ayobami0** [reply]
I was thinking about how to differentiate whether the user is signing in with a username vs email address

**Reply by deelight-del** [reply]
Ooh...
I was proposing that we use the @ symbol to differentiate an email or not. If @ is present then it is an email, otherwise it's a username.
(I implemented this logic for logging in).

If the above logic will stand, then it means that we cannot accept a username that contains @ . When a user tries to register with a username that contains @, we should return a bad request saying something like 'username cannot contain' @...

**Reply by Ayobami0** [reply]
I understand now
